src/providers/ipfs/multi.py

```diff
@@ -0,0 +1,105 @@
+import logging
+from abc import ABC
+from functools import wraps
+from typing import Generic, Iterable, TypeVar, Protocol
+
+from .cid import CIDv0, CIDv1
+from .types import IPFSError, IPFSProvider
+
+logger = logging.getLogger(__name__)
+
+
+T = TypeVar("T")
+
+
+class MultiProvider(Generic[T], ABC):
+    """Base class for working with multiple providers"""
+
+    providers: list[T]
+    current_provider_index: int = 0
+    last_working_provider_index: int = 0
+
+    @property
+    def provider(self) -> T:
+        return self.providers[self.current_provider_index]
+
+
+class SupportsRetries(Protocol):
+    @property
+    def retries(self) -> int:
+        ...
+
+
+class MaxRetryError(IPFSError):
+    ...
+
+
+def retry(fn):
+    @wraps(fn)
+    def wrapped(self: SupportsRetries, *args, **kwargs):
+        retries_left = self.retries
+        while retries_left:
+            try:
+                return fn(self, *args, **kwargs)
+            except IPFSError as ex:
+                retries_left -= 1
+                if not retries_left:
+                    raise MaxRetryError from ex
+                logger.warning({"msg": f"Retrying a failed call of {fn.__name__}, {retries_left=}", "error": str(ex)})
+        raise MaxRetryError
+
+    return wrapped
+
+
+def with_fallback(fn):
+    @wraps(fn)
+    def wrapped(self: MultiProvider, *args, **kwargs):
+        try:
+            result = fn(self, *args, **kwargs)
+        except IPFSError:
+            self.current_provider_index = (self.current_provider_index + 1) % len(self.providers)
+            if self.last_working_provider_index == self.current_provider_index:
+                logger.error({"msg": "No more IPFS providers left to call"})
+                raise
+            return wrapped(self, *args, **kwargs)
+
+        self.last_working_provider_index = self.current_provider_index
+        return result
```

Comment:
When an error occurred, the except branch is being executed, switching the `current_provider_index`. Underlying `provider` property used to make a call is dependent on this index. So that's it.